src/mc-range-encoding.c

```diff
@@ -354,8 +354,9 @@ bool mc_getTypeInfoDouble(mc_getTypeInfoDouble_args_t args,
     if (use_precision_mode) {
         // Take a number of xxxx.ppppp and truncate it xxxx.ppp if precision = 3.
         // We do not change the digits before the decimal place.
-        double v_prime = trunc(args.value * exp10Double(args.precision.value)) / exp10Double(args.precision.value);
-        int64_t v_prime2 = (int64_t)((v_prime - args.min.value) * exp10Double(args.precision.value));
+        int64_t v_prime = (int64_t)(trunc(args.value * exp10Double(args.precision.value)));
+        int64_t scaled_min = (int64_t)(args.min.value * exp10Double(args.precision.value));
+        int64_t v_prime2 = v_prime - scaled_min;
```

Comment:
I thought of another test case where this might fail:
```js
let min = 900719925474099.0; // <-- exactly representable as double
let max = 900719925474100.0; // <-- exactly representable
let val = 900719925474099.6; // <-- exactly representable
let other_val = 900719925474099.5; // <-- exactly representable
let prc = 1;

// multiplying each by 10^prc = 10, it follows that
let scaled_min = 9007199254740990; // <-- exactly representable
let scaled_max = 9007199254741000; // <-- exactly representable
let scaled_val = 9007199254740996; // <-- exactly representable
let scaled_other_val = 9007199254740996; // <-- 9007199254740995 is NOT exactly representable; rounded to +1
```
It follows that both `scaled_val` and `scaled_other_val` return the same `v_prime2` value of `6`,  even though they are unique inputs

Reply:
Both `9007199254740995` and `9007199254740996` are the same number in double (both 9007199254740996). So once the precision calculation kicks in, we lose accuracy by scaling it up from `900719925474099.6` (error -0.025) and `900719925474099.5` (error 0). The precision calculation is always going to have error in it because it `double`. This is why this double math is so frustrating to me because it is "easy" to run into rounding.

See https://flop.evanau.dev/double-converter

Reply:
I still think we need to assert that `scaled_max` does not exceed 2^53 (the rounding threshold), so that we can be sure that the value never runs into rounding when scaled by `10^prc`.

That way this equation will always be true before proceeding: `(value * 10^prc) <= scaled_max <= 2^53`

Reply:
Yea, I'm with Mark on this one. `other_val` cannot be expected to be a different value than `val` unfortunately, even if the range is so small. 

This will however have to be something that we have to be very careful to clarify in the documentation, and in how we teach users to use this feature. It may even be necessary for us to allow users to enable and disable the precision optimization mode, since as Erwin said it's counter intuitive for `900719925474099.5` and `900719925474099.6` to yield different values when they are representable as doubles.

Reply:
Okay.. I thought about this some more, and I think my assertion on `scaled_max` being less than `2^53` is not enough to ensure there are no collisions when mapping two values within the range.  Looking at how floating point multiplication works, I can imagine how it's still possible for two distinct double values, `x` and `y`, to produce the same result when multiplied by a constant (`10^prc`), and for the result to still be less than `2^53`.  At any rate, I don't have the brainwaves right now to be able to prove (or refute) this, or provide an example.. Probably wise to consult the crypt team, though, right?

Reply:
I think the assertion around 2^53 for precison mode is still useful. There are cases though that can produce imprecise results due to rounding even with it. Floating point does have rounding and some numbers cannot be represented precisely. 

When a user types `9007199254740995` or `9007199254740996` in C/C++, they think the numbers are different but they are really the same in doubles. I think we will just need to document to users that setting precision may be imprecise due to floating point rounding.